Expose torch.futures.Future

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39267 Add rpc.async_function decorator for TorchScript functions
* #39216 Add rpc.async_function decorator for rpc_async
* **#39008 Expose torch.futures.Future**

This commit adds a `torch.futures.Future` type and exposes its ctor,
`wait`, `then`, and `set_result` APIs. This type is currently a
wrapper of `c10::ivalue::Future` and mainly used by RPC for now. Later,
we could revamp c10d APIs to return this `Future` type as well. More
utils will be added into `torch.futures` package in followup PRs.

Differential Revision: [D21723022](https://our.internmc.facebook.com/intern/diff/D21723022)

mrshenli committed:
Expose torch.futures.Future

This commit adds a `torch.futures.Future` type and exposes its ctor,
`wait`, `then`, and `set_result` APIs. This type is currently a
wrapper of `c10::ivalue::Future` and mainly used by RPC for now. Later,
we could revamp c10d APIs to return this `Future` type as well. More
utils will be added into `torch.futures` package in followup PRs.

[ghstack-poisoned]

mrshenli committed:
Update on "Expose torch.futures.Future"


This commit adds a `torch.futures.Future` type and exposes its ctor,
`wait`, `then`, and `set_result` APIs. This type is currently a
wrapper of `c10::ivalue::Future` and mainly used by RPC for now. Later,
we could revamp c10d APIs to return this `Future` type as well. More
utils will be added into `torch.futures` package in followup PRs.

[ghstack-poisoned]

mrshenli committed:
Update on "Expose torch.futures.Future"


This commit adds a `torch.futures.Future` type and exposes its ctor,
`wait`, `then`, and `set_result` APIs. This type is currently a
wrapper of `c10::ivalue::Future` and mainly used by RPC for now. Later,
we could revamp c10d APIs to return this `Future` type as well. More
utils will be added into `torch.futures` package in followup PRs.

[ghstack-poisoned]

mrshenli committed:
Update on "Expose torch.futures.Future"


This commit adds a `torch.futures.Future` type and exposes its ctor,
`wait`, `then`, and `set_result` APIs. This type is currently a
wrapper of `c10::ivalue::Future` and mainly used by RPC for now. Later,
we could revamp c10d APIs to return this `Future` type as well. More
utils will be added into `torch.futures` package in followup PRs.

Differential Revision: [D21723022](https://our.internmc.facebook.com/intern/diff/D21723022)

[ghstack-poisoned]

mrshenli committed:
Update on "Expose torch.futures.Future"


This commit adds a `torch.futures.Future` type and exposes its ctor,
`wait`, `then`, and `set_result` APIs. This type is currently a
wrapper of `c10::ivalue::Future` and mainly used by RPC for now. Later,
we could revamp c10d APIs to return this `Future` type as well. More
utils will be added into `torch.futures` package in followup PRs.

Differential Revision: [D21723022](https://our.internmc.facebook.com/intern/diff/D21723022)

[ghstack-poisoned]

mrshenli committed:
Update on "Expose torch.futures.Future"


This commit adds a `torch.futures.Future` type and exposes its ctor,
`wait`, `then`, and `set_result` APIs. This type is currently a
wrapper of `c10::ivalue::Future` and mainly used by RPC for now. Later,
we could revamp c10d APIs to return this `Future` type as well. More
utils will be added into `torch.futures` package in followup PRs.

Differential Revision: [D21723022](https://our.internmc.facebook.com/intern/diff/D21723022)

[ghstack-poisoned]

mrshenli committed:
Update on "Expose torch.futures.Future"


This commit adds a `torch.futures.Future` type and exposes its ctor,
`wait`, `then`, and `set_result` APIs. This type is currently a
wrapper of `c10::ivalue::Future` and mainly used by RPC for now. Later,
we could revamp c10d APIs to return this `Future` type as well. More
utils will be added into `torch.futures` package in followup PRs.

Differential Revision: [D21723022](https://our.internmc.facebook.com/intern/diff/D21723022)

[ghstack-poisoned]

mrshenli committed:
Update on "Expose torch.futures.Future"


This commit adds a `torch.futures.Future` type and exposes its ctor,
`wait`, `then`, and `set_result` APIs. This type is currently a
wrapper of `c10::ivalue::Future` and mainly used by RPC for now. Later,
we could revamp c10d APIs to return this `Future` type as well. More
utils will be added into `torch.futures` package in followup PRs.

Differential Revision: [D21723022](https://our.internmc.facebook.com/intern/diff/D21723022)

[ghstack-poisoned]

rohan-varma commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -259,7 +259,10 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
   void markCompleted(IValue value) {
     std::unique_lock<std::mutex> lock(mutex_);
-    AT_ASSERT(!completed());
+    TORCH_CHECK(
```
Thanks for improving the error message here!

rohan-varma commented on test/test_futures.py:
```diff
@@ -0,0 +1,101 @@
+        f = Future()
+
+        t = threading.Thread(target=slow_set_future, args=(f, torch.ones(2, 2)))
+        t.daemon = True
```
Curious, why set this as a daemon instead of joining after the assertEqual?

mrshenli replied:
it should be joined, my bad.

rohan-varma commented on torch/futures/__init__.py:
```diff
@@ -0,0 +1,89 @@
+    def __new__(cls):
+        return super(Future, cls).__new__(cls)
+
+    def wait(self):
```
Nit: Should we expose `value()`/`constValue()` as well?

mrshenli replied:
Let me add a `result()` API to it.

rohan-varma commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -827,13 +827,34 @@ void initJITBindings(PyObject* module) {
+          "set_result",
+          // Intentionally not releasing GIL
+          &PythonFutureWrapper::markCompleted)
+      .def(
```
Should we use `py::pickle`, as in https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/init.cpp#L295, otherwise I think there will be a bunch of warnings spewed in logs?

mrshenli replied:
Yes, let me update. Thanks!

rohan-varma commented on torch/futures/__init__.py:
```diff
@@ -0,0 +1,89 @@
+            >>>     target=slow_set_future,
+            >>>     args=(fut, torch.ones(2) * 3)
+            >>> )
+            >>> t.daemon = True
```
This can be updated to join as well if that's the intended use pattern

rohan-varma commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -2580,7 +2603,59 @@ def test_callback_none(self):
+            rpc.rpc_async(dst, raise_func).then(None)
+
+    @dist_init
+    @_skip_if_tensorpipe_agent
```
Do we know why this doesn't work for tensorpipe? It seems like this high-level futures API should be agent agnostic. Although, we can file an issue for this and not block this PR

rohan-varma reviewed (approved):
LGTM mod some minor comments.

pritamdamania87 commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -827,13 +827,34 @@ void initJITBindings(PyObject* module) {
           &PythonFutureWrapper::then,
+          py::call_guard<py::gil_scoped_release>())
+      .def(
+          "set_result",
+          // Intentionally not releasing GIL
+          &PythonFutureWrapper::markCompleted)
```
Not as part of this PR, but do we have plans to have a way to set an error on the Future similar to our C++ API?

mrshenli replied:
Yes, that should be part of the Python `Future` API. @jjlilley also mentioned that. Will add an issue to track.

pritamdamania87 commented on torch/futures/__init__.py:
```diff
@@ -0,0 +1,89 @@
+
+
+class Future(torch._C.Future):
+    r"""
+    Wrapper around a ``torch._C.Future``.
+    """
```
Shouldn't we have some details documentation here since this is a user facing API?

mrshenli replied:
Let me expand.

pritamdamania87 commented on torch/futures/__init__.py:
```diff
@@ -0,0 +1,89 @@
+        r"""
+        Block until the value of this ``Future`` is ready.
+
+        Return:
+            The value held by this ``Future``.
```
Should we also mention this could also throw exceptions?

mrshenli replied:
Good point, will add that

mrshenli committed:
Update on "Expose torch.futures.Future"


This commit adds a `torch.futures.Future` type and exposes its ctor,
`wait`, `then`, and `set_result` APIs. This type is currently a
wrapper of `c10::ivalue::Future` and mainly used by RPC for now. Later,
we could revamp c10d APIs to return this `Future` type as well. More
utils will be added into `torch.futures` package in followup PRs.

Differential Revision: [D21723022](https://our.internmc.facebook.com/intern/diff/D21723022)

[ghstack-poisoned]

mrshenli committed:
Update on "Expose torch.futures.Future"


This commit adds a `torch.futures.Future` type and exposes its ctor,
`wait`, `then`, and `set_result` APIs. This type is currently a
wrapper of `c10::ivalue::Future` and mainly used by RPC for now. Later,
we could revamp c10d APIs to return this `Future` type as well. More
utils will be added into `torch.futures` package in followup PRs.

Differential Revision: [D21723022](https://our.internmc.facebook.com/intern/diff/D21723022)

[ghstack-poisoned]

mrshenli committed:
Update on "Expose torch.futures.Future"


This commit adds a `torch.futures.Future` type and exposes its ctor,
`wait`, `then`, and `set_result` APIs. This type is currently a
wrapper of `c10::ivalue::Future` and mainly used by RPC for now. Later,
we could revamp c10d APIs to return this `Future` type as well. More
utils will be added into `torch.futures` package in followup PRs.

Differential Revision: [D21723022](https://our.internmc.facebook.com/intern/diff/D21723022)

[ghstack-poisoned]

mrshenli committed:
Update on "Expose torch.futures.Future"


This commit adds a `torch.futures.Future` type and exposes its ctor,
`wait`, `then`, and `set_result` APIs. This type is currently a
wrapper of `c10::ivalue::Future` and mainly used by RPC for now. Later,
we could revamp c10d APIs to return this `Future` type as well. More
utils will be added into `torch.futures` package in followup PRs.

Differential Revision: [D21723022](https://our.internmc.facebook.com/intern/diff/D21723022)

[ghstack-poisoned]

mrshenli committed:
Update on "Expose torch.futures.Future"


This commit adds a `torch.futures.Future` type and exposes its ctor,
`wait`, `then`, and `set_result` APIs. This type is currently a
wrapper of `c10::ivalue::Future` and mainly used by RPC for now. Later,
we could revamp c10d APIs to return this `Future` type as well. More
utils will be added into `torch.futures` package in followup PRs.

Differential Revision: [D21723022](https://our.internmc.facebook.com/intern/diff/D21723022)

[ghstack-poisoned]
